src/jit/morph.cpp

```diff
@@ -9505,14 +9534,26 @@ GenTreePtr Compiler::fgMorphOneAsgBlockOp(GenTreePtr tree)
                 }
             }
 
-            if (src->OperIsIndir() && !fgIsIndirOfAddrOfLocal(src))
+            if (src->OperIsIndir())
             {
-                // If we have no information about the src, we have to assume it could
-                // live anywhere (not just in the GC heap).
-                // Mark the GT_IND node so that we use the correct write barrier helper in case
-                // the field is a GC ref.
+                if (!fgIsIndirOfAddrOfLocal(src))
+                {
+                    // If we have no information about the src, we have to assume it could
+                    // live anywhere (not just in the GC heap).
+                    // Mark the GT_IND node so that we use the correct write barrier helper in case
+                    // the field is a GC ref.
+                    src->gtFlags |= (GTF_GLOB_REF | GTF_IND_TGTANYWHERE);
+                }
 
-                src->gtFlags |= (GTF_EXCEPT | GTF_GLOB_REF | GTF_IND_TGTANYWHERE);
+                src->gtFlags &= (~GTF_EXCEPT | src->AsIndir()->Addr()->gtFlags);
+                if (src->OperMayThrow(this))
+                {
+                    src->gtFlags |= GTF_EXCEPT;
+                }
+                else
+                {
+                    src->gtFlags |= GTF_IND_NONFAULTING;
+                }
```

Comment:
Not a big deal, but the above pattern appears a lot - it might be nice to extract it into a method.